## Code Review — PR #128

**Revue multi-domaines** : 5 agents frais (`security-auditor`, `architect`, `product-lead`, `tester`, `backend-dev`) · pipeline `/code-review` · 6 fichiers (+213/−3) · CI verte (lint → test → build, typecheck advisory ✅).

### Verdict : 🔴 Request changes (nuancé — voir bas de commentaire)

---

### 🚫 Bloquants

**issue(blocking): [Sécurité — Medium, résiduel pré-existant] Le GET mock conserve un pouvoir de mutation de paiement non authentifié ; la garantie repose sur l'hygiène des variables d'environnement plutôt que sur une capacité vérifiable par requête**
`src/pages/api/stripe-webhook.ts:146` — security-auditor — Confiance : 88 %
Root cause : `netlify dev` démarre `astro dev` (DEV=true) tout en injectant les variables du dashboard Netlify, y compris celles scopées production. Un `GOOGLE_CALENDAR_MOCK=true` oublié dans n'importe quel scope du dashboard ré-ouvre donc la passerelle contre des identifiants de prod sur la machine du développeur. Une fois ouvert, le handler est un GET à effet de bord sans secret par requête ni vérification loopback : un `<img src="http://localhost:4321/api/stripe-webhook/?mock=1&appointment_id=...">` sur n'importe quelle page chargée par le développeur déclenche la mutation (forme CSRF), et `/rdv/merci/?mock=1&appointment_id=...` fait de même côté serveur via le frontmatter. Atténuants : IDs de RDV en UUIDv4 (non énumérables), chaîne de préconditions longue → sévérité Medium, pas High. L'auditeur précise : follow-up de durcissement recommandé, pas une raison de bloquer l'amélioration.
Class: `[missing-input-validation]`
Raw callsites: `src/pages/api/stripe-webhook.ts:146`, `src/pages/api/stripe-webhook.ts:158`, `src/pages/rdv/merci.astro:33`
Solutions :
1. (recommandé) Exiger un token partagé dev-only (`MOCK_WEBHOOK_TOKEN`, jamais défini dans le dashboard Netlify, comparé en temps constant) en plus de la porte DEV + assert `url.hostname` loopback.
2. Retirer le handler GET du code livré — déclencher le mock via un script npm / recette curl contre `astro dev`.

**todo: Les branches restantes du GET ne sont pas testées (400 sans `mock=1`, 400 sans `appointment_id`, 500 si `handlePaymentSucceeded` lève)**
`src/pages/api/stripe-webhook.ts:145-176` — tester — Confiance : 85 %
Root cause : le PR ne couvre que la nouvelle porte DEV ; les chemins de validation de paramètres et d'erreur n'ont jamais eu de test (le GET n'en avait aucun avant ce PR). Endpoint dev-only, sans impact sécurité — lacune pré-existante.
Solutions :
1. (recommandé) Paramétrer `callMockGet(query)` et ajouter 3 tests (400 / 400 / 500), en réutilisant `expectNoPaymentSideEffects` là où applicable.
2. Accepter pour un endpoint dev-only et documenter la lacune dans `aidd_docs/memory/testing.md`.

---

### ⚠️ Avertissements (non bloquants)

**suggestion(non-blocking): Aucun test positif ne prouve que la passerelle s'ouvre encore en dev (DEV=true ∧ MOCK=true) — une régression « toujours fermé » passe la suite entière**
`tests/unit/stripe-webhook.test.ts:143` — tester (95 %), product-lead (90 %), backend-dev (88 %), architect (80 %)
Preuve par mutation **exécutée** : porte forcée en always-deny → les 5 tests du fichier restent verts. La direction « refus » est bien épinglée (suppression du conjonct `DEV` → échec réel `AssertionError: expected 500 to be 403` ; `&&`→`||` → les deux tests échouent) ; la direction « ouverture » ne l'est pas. La claim du PR « le parcours mock local reste disponible » est donc assertée, pas prouvée.
Class: `[candidate/incomplete-branch-coverage]`
Raw callsites: `tests/unit/stripe-webhook.test.ts:153`, `src/pages/api/stripe-webhook.ts:139`
Solutions :
1. (recommandé) 3ᵉ test : `DEV=true` + `GOOGLE_CALENDAR_MOCK=true` avec `mockFirstDeliveryWins()` → assert 200, corps `{ ok: true, mocked: true }`, et `stripe_event_id: 'evt_mock_appt001'` (verrouille aussi la dérivation d'IDs).
2. Variante sans chaîne DB (backend-dev) : mêmes flags mais URL sans `mock` → 400 « Paramètre mock=1 requis » — prouve l'ouverture de la porte sans mocker la chaîne Supabase.

**suggestion(non-blocking): Lecteurs frères de `GOOGLE_CALENDAR_MOCK` sans garde DEV — le même flag a désormais deux sémantiques de sécurité différentes à trois endroits**
`src/lib/google-calendar.ts:114`, `src/lib/calendar-cache.ts:13` — architect (82 %), security-auditor (80 %), backend-dev (78 %), product-lead (thought, 75 %)
Si le flag fuite dans un environnement déployé (scénario racine de #70), la prod sert des créneaux fictifs (mercredis) à de vrais patients et contourne le cache Blobs — impact intégrité/opérationnel, sans bypass d'auth. Couplage sémantique : le mock de paiement Stripe dépend d'un flag nommé « Google Calendar ». Question ouverte associée (backend-dev, 70 %) : vérifier que `GOOGLE_CALENDAR_MOCK` est bien supprimé de **tous** les scopes du dashboard Netlify — la correction code seule ne neutralise pas ces chemins-là.
Class: `[parallel-path-drift]`
Raw callsites: `src/lib/google-calendar.ts:114`, `src/lib/calendar-cache.ts:13`, `src/pages/api/stripe-webhook.ts:139`
Solutions :
1. (recommandé) Follow-up tracké : helper partagé `src/lib/mock-mode.server.ts` (`isCalendarMockEnabled` / `isPaymentMockEnabled`) aux trois sites — ou même conjonct `import.meta.env.DEV &&` sur les deux lecteurs (attention : changement de comportement pour `astro preview`/QA, à décider).
2. Opérationnel : supprimer le flag de tous les scopes Netlify ; documenter dans `docs/LOCAL_DEV.md` qu'il ne vit que dans le `.env` local.

**suggestion(non-blocking): `merci.astro` — la moitié « page » du flux mock n'est pas synchronisée avec la nouvelle porte de la route**
`src/pages/rdv/merci.astro:65-74` — product-lead (85 %), security-auditor (82 %), backend-dev (82 %), architect (70 %)
Le fetch serveur ignore le statut de réponse (le `catch` n'attrape que les erreurs réseau) et n'a pas de timeout ; la variante « payment-success » se déduit uniquement des params d'URL. En prod, `?mock=1` fabrique donc une page « Paiement reçu » sans aucune mutation derrière (affichage seul — page `noindex`, échappement auto, aucun XSS) ; en QA de build (`npm run preview`, DEV=false), échec silencieux déguisé en succès. Comportement pré-existant, mais ce PR transforme le 403 en réponse normale hors dev.
Class: `[parallel-path-drift, missing-error-handling, unbounded-loop]`
Raw callsites: `src/pages/rdv/merci.astro:65`, `src/pages/rdv/merci.astro:74`
Solutions :
1. (recommandé) Gâter la variante et le fetch sur `import.meta.env.DEV` aussi (même prédicat partagé que la route), vérifier `res.ok`, ajouter `signal: AbortSignal.timeout(3000)`.
2. Minime : timeout + `if (!res.ok) logger.warn(...)`, et retirer `isMockPayment` du prédicat de variante.

**suggestion(non-blocking): Rien n'épingle l'invariant au niveau build (le constant-folding de `DEV` dans le bundle SSR Netlify)**
`tests/unit/stripe-webhook.test.ts:129` — security-auditor — Confiance : 85 %
Les tests stubbent `import.meta.env` à l'exécution dans Vitest ; une régression de config Vite/Astro (ex. override `define`) rouvrirait la porte avec 167/167 tests verts.
Solutions :
1. (recommandé) Étape CI : `npm run build` + serveur preview + `GET /api/stripe-webhook/?mock=1&appointment_id=x` → assert 403.
2. Grep statique du conjonct inliné dans le bundle SSR émis (zéro coût runtime, plus fragile).

---

### 💡 Suggestions mineures

- **suggestion(non-blocking)** : handler mock co-localisé avec le contrat Stripe de prod dans un module de ~380 lignes, sans élagage build-time — architect (72 %). Vite constant-fold la porte en prod (risque résiduel = cohésion/distance d'édition, pas un chemin runtime vivant). Solu : `export const GET = import.meta.env.DEV ? mockGet : rejectMock` (DCE en prod) — ou accepter avec un commentaire « dev-only fixture — do not weaken ».
- **suggestion(non-blocking)** : périmètre des docs `aidd_docs/` — architect (75 %) + product-lead (80 %, nitpick) : `testing.md` embarque un runbook WSL/fnm spécifique à une machine dans la mémoire projet auto-chargée (→ `docs/LOCAL_DEV.md` ou `memory/internal/`), `coding-assertions.md` devient une 3ᵉ source de vérité pour les gates qualité (marquer AGENTS.md comme canonique), et ces 4 fichiers ne sont pas mentionnés dans la description du PR.
- **nitpick** : hygiène de test — stubs d'env au niveau fichier sans `vi.unstubAllEnvs()` (tester, 90 %) ; cast `as Parameters<typeof GET>[0]` (tester, 85 % — bénin, vérifié : GET ne déstructure que `url`) ; corps 403/400 en texte brut vs JSON pour 200/500 (backend-dev, 90 % — recommande laisser tel quel, cohérent avec le POST).

---

### ✅ Praise

- **praise** : le mécanisme est le bon, **vérifié dans les sources de Vite 6** : `import.meta.env.DEV` est inliné statiquement (`DEV: !isProduction`) dans tout artefact `astro build`, bundle SSR Netlify inclus — la conjonction se constant-fold en `false`, impossible à ré-ouvrir au runtime, contrairement à l'ancienne porte (clés custom non inlinées, lues dans `process.env` à l'exécution). — backend-dev (95 %)
- **praise** : tests honnêtes, **prouvés par mutation exécutée** (pas seulement CI verte) : suppression du conjonct DEV → échec réel ; `&&`→`||` → les deux tests échouent ; les assertions « pas d'effet de bord » ciblent de vrais `vi.fn` de modules mockés (`vi.stubEnv('DEV', boolean)` est une overload typée de première classe en Vitest 4.1.9). — tester (97 %)
- **praise** : les 5 critères d'acceptation du spec interne sont remplis avec preuves ; zéro suppression TypeScript (l'import typé `Stripe` est intact). — product-lead (95 %)
- **praise** : le refactoring d'import hoisté dans le test L2 est équivalent en comportement (même instance de module mockée) et plus propre. — tester (93 %)
- **praise** : les 4 fichiers `aidd_docs/` remplissent des slots pré-déclarés dans AGENTS.md (2 des 16 includes `@aidd_docs/memory/*` étaient dangling sur `main`). — architect (90 %)
- **praise** : scan secrets des 4 nouveaux docs : propre ; la vérification de signature du POST est intacte ; le 403 précède tout parsing de params (les indices 400 ne fuient pas en prod). — security-auditor

---

### Verdict

🔴 **Request changes** — au sens strict des étiquettes : un `issue:` (résiduel sécurité Medium, **pré-existant** et rétréci par ce PR) et un `todo:` (branches 400/500 non testées, **pré-existantes**).

**Le changement de code lui-même est correct, minimal, et salué par les cinq domaines** — il ferme bien la cause racine de #70 (fuite de flag dashboard ne pouvant plus ouvrir la route en prod). Chemin le plus court vers le merge : ajouter le test positif d'ouverture (+ idéalement les tests 400/500 dans le même describe paramétré) et trancher le résiduel sécurité en follow-up tracké (token dev-only ou retrait du handler GET) — ou merger en connaissance de cause avec les follow-ups.

--- a/aidd_docs/tasks/2026_08/2026_08_28_secure_mock_webhook/phase-1.md
+++ b/aidd_docs/tasks/2026_08/2026_08_28_secure_mock_webhook/phase-1.md
@@ -1,0 +1,67 @@
+---
+status: done
+---
+
+# Instruction: Gate and verify the mock webhook
+
+## Architecture projection
+
+> Tree of the final files. ✅ create · ✏️ modify · ❌ delete
+
+```txt
+.
+├── src/pages/api/stripe-webhook.ts       ✏️ require development mode for mock GET access
+└── tests/unit/stripe-webhook.test.ts    ✏️ cover the environment gate without running payment side effects
+```
+
+## User Journey
+
+```mermaid
+flowchart TD
+  A[Caller requests mock webhook GET] --> B{DEV and GOOGLE_CALENDAR_MOCK enabled?}
+  B -- No --> C[Return HTTP 403 without payment side effects]
+  B -- Yes --> D[Continue existing mock validation and payment flow]
+```
+
+## Test Scope
+
+```mermaid
+---
+title: Test scope
+---
+journey
+  section Setup
+    Mock webhook dependencies and environment => isolated route ready: 5: api
+  section Happy path
+    Disable development mode and enable calendar mock => GET request returns 403 without appointment mutation: 5: api
+  section Edge case - Calendar mock disabled
+    Enable development mode and disable calendar mock => GET request returns 403 without appointment mutation: 1: api
+```
+
+## Tasks to do
+
+### `1)` Enforce the development-only gate
+
+> Reject the mock GET route unless both required environment conditions are true.
+
+1. Combine `import.meta.env.DEV` with `GOOGLE_CALENDAR_MOCK` at the route boundary.
+2. Preserve the existing development mock flow and French API responses.
+3. Confirm the stale Stripe suppression is absent and keep the typed Stripe import.
+
+### `2)` Prove the route boundary
+
+> Add focused regression coverage and run the project validation commands.
+
+1. Exercise both disabled-environment branches through the exported GET handler.
+2. Assert payment persistence and notification side effects are not invoked.
+3. Run the focused test, lint, typecheck, test suite, and production build.
+
+## Test acceptance criteria
+
+| Task | Acceptance criteria |
+| ---- | ------------------- |
+| 1 | A request is rejected with HTTP 403 whenever development mode is false, even if the calendar mock flag is true. |
+| 1 | A request is rejected with HTTP 403 whenever the calendar mock flag is false, even if development mode is true. |
+| 1 | The Stripe namespace import resolves without a TypeScript suppression. |
+| 2 | Rejected requests cause no appointment update, email, or calendar event. |
+| 2 | Existing Stripe webhook unit tests, lint, typecheck, and production build complete successfully. |

--- a/aidd_docs/tasks/2026_08/2026_08_28_secure_mock_webhook/plan.md
+++ b/aidd_docs/tasks/2026_08/2026_08_28_secure_mock_webhook/plan.md
@@ -1,0 +1,19 @@
+---
+objective: "The Stripe mock webhook GET route is unusable outside Vite development mode and remains type-safe."
+status: in-progress
+---
+
+# Plan: Secure the Stripe mock webhook
+
+## Overview
+
+| Field      | Value |
+| ---------- | ----- |
+| **Goal**   | Restrict the unauthenticated mock payment endpoint to explicit local development mode. |
+| **Source** | GitHub issue [#70](https://github.com/tavianm/omf-therapie/issues/70) |
+
+## Phases
+
+| #   | Phase | File |
+| --- | ----- | ---- |
+| 1   | Gate and verify the mock webhook | [`phase-1.md`](./phase-1.md) |

--- a/aidd_docs/tasks/2026_08/2026_08_28_secure_mock_webhook/plan.md
+++ b/aidd_docs/tasks/2026_08/2026_08_28_secure_mock_webhook/plan.md
@@ -1,6 +1,6 @@
 ---
 objective: "The Stripe mock webhook GET route is unusable outside Vite development mode and remains type-safe."
-status: in-progress
+status: implemented
 ---
 
 # Plan: Secure the Stripe mock webhook

--- a/aidd_docs/tasks/2026_08/2026_08_28_secure_mock_webhook/plan.md
+++ b/aidd_docs/tasks/2026_08/2026_08_28_secure_mock_webhook/plan.md
@@ -1,6 +1,6 @@
 ---
 objective: "The Stripe mock webhook GET route is unusable outside Vite development mode and remains type-safe."
-status: implemented
+status: reviewed
 ---
 
 # Plan: Secure the Stripe mock webhook

--- a/src/pages/api/stripe-webhook.ts
+++ b/src/pages/api/stripe-webhook.ts
@@ -141,7 +141,8 @@ export const POST: APIRoute = async ({ request }) => {
 };
 
 export const GET: APIRoute = async ({ url }) => {
-  const isMockMode = import.meta.env.GOOGLE_CALENDAR_MOCK === 'true';
+  const isMockMode =
+    import.meta.env.DEV && import.meta.env.GOOGLE_CALENDAR_MOCK === 'true';
   if (!isMockMode) {
     return new Response('Mode mock non activé', { status: 403 });
   }

--- a/tests/unit/stripe-webhook.test.ts
+++ b/tests/unit/stripe-webhook.test.ts
@@ -57,7 +57,9 @@ vi.mock('@/lib/stripe', () => ({ getStripe: () => null }));
 
 // --- Import after mocks -----------------------------------------------------
 
-import { handlePaymentSucceeded } from '@/pages/api/stripe-webhook';
+import { GET, handlePaymentSucceeded } from '@/pages/api/stripe-webhook';
+import { supabaseAdmin } from '@/lib/supabase';
+import { createCalendarEvent } from '@/lib/google-calendar';
 
 // --- Helpers ----------------------------------------------------------------
 
@@ -108,6 +110,8 @@ function mockRetryReRead(apptOverrides: Partial<Record<string, unknown>> = {}) {
 
 beforeEach(() => {
   vi.clearAllMocks();
+  vi.stubEnv('DEV', true);
+  vi.stubEnv('GOOGLE_CALENDAR_MOCK', 'false');
   // Default: the mark-delivered UPDATE succeeds.
   mockSupabaseChain.update.mockReturnThis();
   mockSupabaseChain.eq.mockReturnThis();
@@ -115,6 +119,46 @@ beforeEach(() => {
   mockSupabaseChain.select.mockReturnThis();
   // Default email result: success.
   mockBuildAndSend.mockResolvedValue({ patientEmailSent: true, therapistEmailSent: true });
+});
+
+// ---------------------------------------------------------------------------
+// Mock GET route — available only in explicit local development mode
+// ---------------------------------------------------------------------------
+
+describe('GET — development mock gate', () => {
+  async function callMockGet(): Promise<Response> {
+    return GET({
+      url: new URL(
+        'http://localhost/api/stripe-webhook/?mock=1&appointment_id=appt_001',
+      ),
+    } as Parameters<typeof GET>[0]);
+  }
+
+  function expectNoPaymentSideEffects(): void {
+    expect(supabaseAdmin.from).not.toHaveBeenCalled();
+    expect(mockBuildAndSend).not.toHaveBeenCalled();
+    expect(createCalendarEvent).not.toHaveBeenCalled();
+  }
+
+  it('returns 403 outside development even when calendar mock mode is enabled', async () => {
+    vi.stubEnv('DEV', false);
+    vi.stubEnv('GOOGLE_CALENDAR_MOCK', 'true');
+
+    const response = await callMockGet();
+
+    expect(response.status).toBe(403);
+    expectNoPaymentSideEffects();
+  });
+
+  it('returns 403 in development when calendar mock mode is disabled', async () => {
+    vi.stubEnv('DEV', true);
+    vi.stubEnv('GOOGLE_CALENDAR_MOCK', 'false');
+
+    const response = await callMockGet();
+
+    expect(response.status).toBe(403);
+    expectNoPaymentSideEffects();
+  });
 });
 
 // ---------------------------------------------------------------------------
@@ -131,7 +175,7 @@ describe('handlePaymentSucceeded — L2 gate (confirmation_sent_at)', () => {
 
     // Assert — NO email sent, NO calendar event created, NO mark-delivered UPDATE.
     expect(mockBuildAndSend).not.toHaveBeenCalled();
-    expect(vi.mocked(await import('@/lib/google-calendar')).createCalendarEvent).not.toHaveBeenCalled();
+    expect(createCalendarEvent).not.toHaveBeenCalled();
     // The only UPDATEs should be the N1 (which failed) — no mark-delivered.
     // We check that .update was not called with confirmation_sent_at by verifying
     // buildAndSend was never called (the side effect that precedes the mark).

